File: ios/LookSeeProto/UI Files/Library.swift
import SwiftUI

struct Library: View {
    @StateObject private var libraryService = LibraryService.shared
    @StateObject private var modelService = ModelService.shared
    @State private var searchText = ""

    var body: some View {
        NavigationStack {
            Group {
                if libraryService.isLoading {
                    VStack(spacing: 16) {
                        ProgressView()
                        Text("Loading landmarks…")
                            .font(.footnote)
                            .foregroundStyle(.secondary)
                    }
                    .frame(maxWidth: .infinity, maxHeight: .infinity)

                } else if let error = libraryService.errorMessage {
                    VStack(spacing: 12) {
                        Image(systemName: "exclamationmark.triangle")
                            .font(.system(size: 32))
                            .foregroundStyle(.orange)
                        Text("Could not load landmarks")
                            .font(.headline)
                        Text(error)
                            .font(.footnote)
                            .foregroundStyle(.secondary)
                            .multilineTextAlignment(.center)
                            .padding(.horizontal, 32)
                        Button("Retry") {
                            Task { await loadForCurrentModel() }
                        }
                        .buttonStyle(.bordered)
                    }
                    .frame(maxWidth: .infinity, maxHeight: .infinity)

                } else if libraryService.items.isEmpty {
                    VStack(spacing: 12) {
                        Image(systemName: "archivebox")
                            .font(.system(size: 32))
                            .foregroundStyle(.secondary)
                        Text("No landmarks found")
                            .font(.headline)
                        Text("No landmarks are associated with your current model.")
                            .font(.footnote)
                            .foregroundStyle(.secondary)
                            .multilineTextAlignment(.center)
                            .padding(.horizontal, 32)
                    }
                    .frame(maxWidth: .infinity, maxHeight: .infinity)

                } else {
                    List(filtered) { item in
                        VStack(alignment: .leading, spacing: 4) {
                            Text(item.label)
                                .font(.headline)
                            Text(item.shortDescription)
                                .font(.footnote)
                                .foregroundStyle(.secondary)
                            Text("Cluster \(item.clusterId)")
                                .font(.caption2)
                                .foregroundStyle(.tertiary)
                        }
                        .padding(.vertical, 4)
                    }
                    .searchable(text: $searchText)
                }
            }
            .navigationTitle("Library")
            .task {
                await loadForCurrentModel()
            }
        }
    }

    // MARK: - Filter
    var filtered: [LibraryLandmark] {
        if searchText.isEmpty {
            return libraryService.items
        }
        return libraryService.items.filter {
            $0.label.localizedCaseInsensitiveContains(searchText) ||
            $0.shortDescription.localizedCaseInsensitiveContains(searchText)
        }
    }

    // MARK: - Load for current model
    private func loadForCurrentModel() async {
        if case .loaded(let infos) = modelService.state,
           let first = infos.first,
           let clusterIdInt = Int(first.name.components(separatedBy: "-").last ?? "") {
            await libraryService.fetchLandmarks(clusterId: clusterIdInt)
        } else {
            libraryService.errorMessage = "No model loaded. Load a model first in Settings."
        }
    }
}

#Preview {
    Library()
}

File: Android/LookSee/app/src/main/java/looksee/angelll/com/uifiles/Library.kt
package looksee.angelll.com.uifiles

import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Archive
import androidx.compose.material.icons.filled.Search
import androidx.compose.material.icons.filled.WarningAmber
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.compose.runtime.collectAsState
import looksee.angelll.com.models.*
import looksee.angelll.com.viewmodels.*
import androidx.compose.runtime.getValue
import androidx.compose.runtime.setValue
import androidx.compose.ui.platform.LocalContext
import kotlinx.coroutines.launch

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun LibraryScreen() {
    val context = LocalContext.current
    val libraryService = remember { LibraryService.shared(context) }
    val modelService = remember { ModelService.shared(context) }
    val coroutineScope = rememberCoroutineScope()

    var searchText by remember { mutableStateOf("") }
    val items by libraryService.items.collectAsState()
    val isLoading by libraryService.isLoading.collectAsState()
    val errorMessage by libraryService.errorMessage.collectAsState()

    val filteredItems = remember(searchText, items) {
        if (searchText.isEmpty()) {
            items
        } else {
            items.filter {
                it.label.contains(searchText, ignoreCase = true) ||
                        it.shortDescription.contains(searchText, ignoreCase = true)
            }
        }
    }

    fun loadForCurrentModel() {
        coroutineScope.launch {
            val state = modelService.state.value
            if (state is ModelState.Loaded) {
                val first = state.models.firstOrNull()
                val clusterIdString = first?.clusterId ?: ""
                val clusterIdInt = clusterIdString.toIntOrNull()

                if (clusterIdInt != null) {
                    libraryService.fetchLandmarks(clusterIdInt)
                }
            }
        }
    }

    LaunchedEffect(Unit) {
        loadForCurrentModel()
    }

    Scaffold(
        topBar = {
            TopAppBar(
                title = { Text("Library", fontWeight = FontWeight.Bold) },
                colors = TopAppBarDefaults.topAppBarColors(
                    containerColor = MaterialTheme.colorScheme.background
                )
            )
        }
    ) { paddingValues ->
        Box(
            modifier = Modifier
                .fillMaxSize()
                .padding(paddingValues)
        ) {
            when {
                isLoading -> {
                    Column(
                        modifier = Modifier.fillMaxSize(),
                        verticalArrangement = Arrangement.Center,
                        horizontalAlignment = Alignment.CenterHorizontally
                    ) {
                        CircularProgressIndicator()
                        Spacer(modifier = Modifier.height(16.dp))
                        Text(
                            text = "Loading landmarks…",
                            fontSize = 13.sp,
                            color = Color.Gray
                        )
                    }
                }
                errorMessage != null -> {
                    Column(
                        modifier = Modifier
                            .fillMaxSize()
                            .padding(horizontal = 32.dp),
                        verticalArrangement = Arrangement.Center,
                        horizontalAlignment = Alignment.CenterHorizontally
                    ) {
                        Icon(
                            imageVector = Icons.Default.WarningAmber,
                            contentDescription = "Error",
                            tint = Color(0xFFFFA500),
                            modifier = Modifier.size(48.dp)
                        )
                        Spacer(modifier = Modifier.height(12.dp))
                        Text(
                            text = "Could not load landmarks",
                            fontWeight = FontWeight.Bold,
                            fontSize = 18.sp
                        )
                        Text(
                            text = errorMessage!!,
                            fontSize = 13.sp,
                            color = Color.Gray,
                            textAlign = TextAlign.Center
                        )
                        Spacer(modifier = Modifier.height(16.dp))
                        Button(onClick = { loadForCurrentModel() }) {
                            Text("Retry")
                        }
                    }
                }
                items.isEmpty() -> {
                    Column(
                        modifier = Modifier
                            .fillMaxSize()
                            .padding(horizontal = 32.dp),
                        verticalArrangement = Arrangement.Center,
                        horizontalAlignment = Alignment.CenterHorizontally
                    ) {
                        Icon(
                            imageVector = Icons.Default.Archive,
                            contentDescription = "Empty",
                            tint = Color.Gray,
                            modifier = Modifier.size(48.dp)
                        )
                        Spacer(modifier = Modifier.height(12.dp))
                        Text(
                            text = "No landmarks found",
                            fontWeight = FontWeight.Bold,
                            fontSize = 18.sp
                        )
                        Text(
                            text = "No landmarks are associated with your current model.",
                            fontSize = 13.sp,
                            color = Color.Gray,
                            textAlign = TextAlign.Center
                        )
                    }
                }
                else -> {
                    Column(modifier = Modifier.fillMaxSize()) {
                        // Search Bar equivalent
                        OutlinedTextField(
                            value = searchText,
                            onValueChange = { searchText = it },
                            placeholder = { Text("Search landmarks") },
                            leadingIcon = { Icon(Icons.Default.Search, contentDescription = null) },
                            modifier = Modifier
                                .fillMaxWidth()
                                .padding(horizontal = 16.dp, vertical = 8.dp),
                            shape = RoundedCornerShape(12.dp),
                            singleLine = true
                        )

                        LazyColumn(
                            modifier = Modifier.fillMaxSize(),
                            contentPadding = PaddingValues(horizontal = 16.dp, vertical = 8.dp)
                        ) {
                            items(filteredItems) { item ->
                                Column(
                                    modifier = Modifier
                                        .fillMaxWidth()
                                        .padding(vertical = 8.dp)
                                ) {
                                    Text(
                                        text = item.label,
                                        fontWeight = FontWeight.Bold,
                                        fontSize = 16.sp
                                    )
                                    Spacer(modifier = Modifier.height(4.dp))
                                    Text(
                                        text = item.shortDescription,
                                        fontSize = 13.sp,
                                        color = Color.Gray
                                    )
                                    Spacer(modifier = Modifier.height(2.dp))
                                    Text(
                                        text = "Cluster ${item.clusterId}",
                                        fontSize = 11.sp,
                                        color = Color.Gray.copy(alpha = 0.7f)
                                    )
                                }
                                HorizontalDivider(color = Color.LightGray.copy(alpha = 0.3f))
                            }
                        }
                    }
                }
            }
        }
    }
}